src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1010,77 +853,370 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                foreach (var member in symbol.GetMembers())
                 {
-                    if (string.IsNullOrWhiteSpace(assembly)) { continue; }
-                    string resolvedAssemblyPath = ResolveAssemblyName(assembly, true);
+                    member.Accept(this);
+                }
+            }
 
-                    // Ignore some specified reference assemblies
-                    string fileName = PathType.GetFileName(resolvedAssemblyPath);
-                    if (s_refAssemblyNamesToIgnore.Value.Contains(fileName))
+            public override void VisitNamedType(INamedTypeSymbol symbol)
+            {
+                // It is namespace-fully-qualified name
+                var symbolFullName = symbol.ToString();
+
+                if (s_sourceTypesCache.TryGetValue(symbolFullName, out int hash))
+                {
+                    if (hash == _hash)
                     {
-                        WriteVerbose(StringUtil.Format(AddTypeStrings.ReferenceAssemblyIgnored, resolvedAssemblyPath));
-                        continue;
+                        DuplicateSymbols.Add(symbolFullName);
                     }
-                    tempReferences.Add(MetadataReference.CreateFromFile(resolvedAssemblyPath));
                 }
-                references = tempReferences.ToArray();
+                else
+                {
+                    UniqueSymbols.Add(symbolFullName);
+                }
+            }
+        }
+
+        private void CacheNewTypes(ConcurrentBag<String> newTypes)
+        {
+            foreach (var typeName in newTypes)
+            {
+                s_sourceTypesCache.Add(typeName, _syntaxTreesHash);
             }
+        }
 
-            CSharpCompilation compilation = CSharpCompilation.Create(
-                PathType.GetRandomFileName(),
-                syntaxTrees: new[] { syntaxTree },
-                references: references,
-                options: new CSharpCompilationOptions(OutputAssemblyTypeToOutputKind(OutputType)));
+        private void CacheAssemply(Assembly assembly)
+        {
+            s_sourceAssemblyCache.Add(_syntaxTreesHash, assembly);
+        }
 
+        private void DoEmitAndLoadAssemply(Compilation compilation, EmitOptions emitOptions)
+        {
             EmitResult emitResult;
 
+            CheckDuplicateTypes(compilation, out ConcurrentBag<String> newTypes);
+
             if (InMemory)
             {
                 using (var ms = new MemoryStream())
                 {
-                    emitResult = compilation.Emit(ms);
+                    emitResult = compilation.Emit(peStream: ms, options: emitOptions);
+
+                    HandleCompilerErrors(emitResult.Diagnostics);
+
                     if (emitResult.Success)
                     {
+                        // TODO:  We could use Assembly.LoadFromStream() in future.
+                        // See https://github.com/dotnet/corefx/issues/26994
                         ms.Flush();
```

Comment:
`Flush()` on `MemoryStream` does nothing, see: http://source.dot.net/#System.Private.CoreLib/shared/System/IO/MemoryStream.cs,179

Reply:
I think we should not rely on the knowledge of the current implementation - what if it changes tomorrow?
We'll remove this in Assembly.LoadFromStream() time.

Reply:
As long as you are using a memory stream, `flush` won't do anything. It's by design because there is no place you need to flush the content to.

Reply:
Removed.